aten/src/ATen/native/cuda/AtomicOps.cuh

```diff
@@ -0,0 +1,344 @@
+  }
+
+  template <typename T>
+  inline __device__ void operator() (c10::complex<T> * address, c10::complex<T> val) {
+    operator()(&address->real_, val.real_);
+    operator()(&address->imag_, val.imag_);
+  }
+};
+
+template <>
+struct gpuAtomic<mul_op> {
+  inline __device__ void operator() (uint8_t * address, uint8_t val) {
+    AtomicIntegerop<uint8_t, sizeof(uint8_t)>()(address, val,
+                                                [](uint8_t val, uint32_t old_byte) {
+                                                  return static_cast<uint8_t>(THCNumerics<uint8_t>::mul(val, old_byte));
+                                                });
+  }
+
+  inline __device__ void operator() (int8_t * address, int8_t val) {
+    AtomicIntegerop<int8_t, sizeof(int8_t)>()(address, val,
+                                              [](int8_t val, uint32_t old_byte) {
+                                                return static_cast<int8_t>(THCNumerics<int8_t>::mul(val, old_byte));
+                                              });
+  }
+
+  inline __device__ void operator() (int16_t * address, int16_t val) {
+    AtomicIntegerop<int16_t, sizeof(int16_t)>()(address, val,
+                                                [](int16_t val, uint32_t old_byte) {
+                                                  return static_cast<uint16_t>(THCNumerics<int16_t>::mul(val, old_byte));
+                                                });    
+  }
+
+  inline __device__ void operator() (int32_t * address, int32_t val) {
+    AtomicIntegerop<int32_t, sizeof(int32_t)>()(address, val,
+                                                [](int32_t val, uint32_t old) {
+                                                  return val * (int32_t)old;
+                                                });
+  }
+
+  inline __device__ void operator() (int64_t * address, int64_t val) {
+    AtomicIntegerop<int64_t, sizeof(int64_t)>()(address, val,
+                                                [](int64_t val, unsigned long long old) {
+                                                  return val * (int64_t)old;
+                                                });    
+  }
+
+  inline __device__ at::Half operator() (at::Half * address, at::Half val) {
+    return AtomicFPOp<at::Half>()(address, val,
+                                  [](at::Half bsum, at::Half val) {
+                                    return THCNumerics<at::Half>::add(bsum, val);
+                                  });
+  }
+
+  inline __device__ at::BFloat16 operator() (at::BFloat16 * address, at::BFloat16 val) {
+    return AtomicFPOp<at::BFloat16>()(address, val,
+                                      [](at::BFloat16 bsum, at::BFloat16 val) {
+                                        return THCNumerics<at::BFloat16>::add(bsum, val);
+                                      });    
+  }
+
+  inline __device__ void operator() (bool * address, bool val) {
+    *address = address && val;
```

Comment:
also does not look right, and is not needed as we are not dispatching bool

Reply:
An implementation for bool has to exist in order to make the `cuda_scatter_gather_base_kernel` in the ScatterGatherKernel.cu file generic enough to work with assignment and mul/add.

Reply:
I don't see why - cuda_scatter_gather_base_kernel should not be dispatched for bool.

Reply:
It uses the same code for the assignment operator as well as add/mult. Boolean assignment has been around before too.